[Neil Banerjee] iP

NeilBaner opened the pull request.

NeilBaner committed:
Level-0

NeilBaner committed:
Start the project

NeilBaner committed:
Merge remote-tracking branch 'origin/master'

NeilBaner committed:
Add echo, greet, goodbye

Add user input, ability to echo the input, ability to quit

NeilBaner committed:
Create Task and TaskList

Create classes for Task and TaskList to be implemented

NeilBaner committed:
Add Done property of Task

NeilBaner committed:
Add more methods to TaskList

Add ability to get list size, mark a task as done, get a specific task. Modify delete to throw exception to be handled by main.

NeilBaner committed:
Implement Mark As Done (basic)

Implement a basic form of the Mark As Done requirement, to be improved.

NeilBaner committed:
Implement Add, List

Add the ability to add tasks and view a list of tasks

NeilBaner committed:
Implement Mark As Done

Add a basic implementation of mark as done, to add more features later.

NeilBaner committed:
Merge remote-tracking branch 'origin/master'

NeilBaner committed:
Improve formatting

Improve both code formatting plus output formatting.

NeilBaner committed:
Create subclasses for different types of tasks

Create subclasses ToDo, Event and Deadline for the three different types of tasks for level 4.

NeilBaner committed:
Add print() method to Task and subclasses

Add a print() method to Task and override it for each of the subclasses, to implement polymorphism so that we can easily print a list of different types of Tasks

NeilBaner committed:
Reimplement list with Task.print(), add ToDo type

Reimplemented the list function using the newly created Task.print() and the subclasses. Much cleaner now. Also rewrote the basic add task function to be a To-Do item instead.

NeilBaner committed:
Implement Event and Deadline

Add the ability to add Event and Deadline type tasks as required for Level 4.

NeilBaner committed:
Fixed output formatting

Fixed some errant spaces in the print output

NeilBaner committed:
Add Messages class

Added a Messages class to print messages, implementing SLAP to some extent.

NeilBaner committed:
Fixed minor formatting

t170815518 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+        this.dueDate = dueDate;
+    }
+
+    @Override
+    public void print() {
+        System.out.print("[D]");
+        super.print();
+        System.out.print(", by " + dueDate);
+        System.out.println();
+    }
```
Is there any reason for you using `print()` instead of overriding `toString()`?

NeilBaner replied:
Yes, there is, I completely forgot about toString() existing :P Whoops, I should change that. Thanks for pointing it out.

t170815518 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+
+    public static void printTaskList() {
+        Messages.printHorizontalLine();
+        if (list.getSize() == 0) {
+            System.out.println("No tasks added yet");
```
I like your treatment for this special case.

t170815518 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+        Messages.printHello();
+        while (toContinue) {
+            input = k.nextLine();
+            if (input.toLowerCase().equals("bye")) {
+                toContinue = false;
+                Messages.printGoodBye();
+            } else if (input.toLowerCase().equals("list")) {
+                printTaskList();
+            } else if (input.toLowerCase().startsWith("done")) {
+                try {
+                    int indexToMark = Integer.parseInt(input.substring(5)) - 1;
+                    markDone(indexToMark);
+                } catch (NumberFormatException e) {
+                    Messages.printFormattingError();
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printNonExistError();
+                }
+            } else if (input.toLowerCase().startsWith("todo")) {
+                addToDo(input.substring(5));
+            } else if (input.toLowerCase().startsWith("deadline")) {
+                try {
+                    String title = input.substring(9, input.indexOf("/by")).strip();
+                    String dueDate = input.substring(input.indexOf("/by") + 4).strip();
+                    addDeadline(title, dueDate);
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printFormattingError();
+                }
+            } else if (input.startsWith("event")) {
+                try {
+                    String title = input.substring(6, input.indexOf("/at")).strip();
+                    String eventTime = input.substring(input.indexOf("/at") + 4).strip();
+                    addEvent(title, eventTime);
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printFormattingError();
+                }
```
Maybe this part could be refactored to a sub-method?

NeilBaner replied:
Good point. Thanks, I'll be sure to do that. SLAP is love, SLAP is life.

t170815518 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printNonExistError();
+                }
+            } else if (input.toLowerCase().startsWith("todo")) {
```
I like the `startsWith()`, which is a neat method to detect the command

t170815518 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,21 @@
+public class Event extends Task {
+    private String eventTime;
```
I would set a getter for attributes of subclasses of Task :)

t170815518 commented on src/main/java/Messages.java:
```diff
@@ -0,0 +1,53 @@
+public class Messages {
```
I like your way to refactor the messages :)

t170815518 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.ArrayList;
+
+public class TaskList {
```
I like how you treat `TaskList`

t170815518 reviewed (commented):
I like your coding style, with sufficient refactoring such as `Message` and `TaskList` Class. I think adding `toString()` would help make the interaction between classes neater.

WYing333 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+        this.dueDate = dueDate;
+    }
+
+    @Override
+    public void print() {
+        System.out.print("[D]");
+        super.print();
+        System.out.print(", by " + dueDate);
+        System.out.println();
+    }
```
Maybe toString() is a more suitable method than print().

NeilBaner replied:
Noted. I completely forgot about the existence of toString(), or I would have used that. Will change asap.

WYing333 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+                    Messages.printNonExistError();
+                }
+            } else if (input.toLowerCase().startsWith("todo")) {
+                addToDo(input.substring(5));
+            } else if (input.toLowerCase().startsWith("deadline")) {
+                try {
+                    String title = input.substring(9, input.indexOf("/by")).strip();
+                    String dueDate = input.substring(input.indexOf("/by") + 4).strip();
+                    addDeadline(title, dueDate);
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printFormattingError();
+                }
+            } else if (input.startsWith("event")) {
+                try {
+                    String title = input.substring(6, input.indexOf("/at")).strip();
+                    String eventTime = input.substring(input.indexOf("/at") + 4).strip();
+                    addEvent(title, eventTime);
+                } catch (IndexOutOfBoundsException e) {
+                    Messages.printFormattingError();
+                }
```
Maybe these addSomething can be refacored to make these shorter. But the naming style is easily understood by others.

WYing333 commented on src/main/java/Messages.java:
```diff
@@ -0,0 +1,53 @@
+        printHorizontalLine();
+        System.out.println("Bye. Hope to see you again soon!");
+        System.out.println("Shutting down...");
+        printHorizontalLine();
+    }
+
+    public static void printAddedToDo(String title) {
+        printHorizontalLine();
+        System.out.println("Done. I've added the following to-do to your list: " + title);
+        printHorizontalLine();
+    }
+
+    public static void printAddedEvent(String title) {
+        Messages.printHorizontalLine();
+        System.out.println("Roger dodger chief. I've added this event to your list: " + title);
+        Messages.printHorizontalLine();
+    }
+
+    public static void printAddedDeadline(String title) {
+        Messages.printHorizontalLine();
+        System.out.println("You got it. Added a new deadline to your list, titled : " + title);
+        Messages.printHorizontalLine();
+    }
+
+    public static void printMarkedDone(String title) {
+        Messages.printHorizontalLine();
+        System.out.println(title + " marked as done.");
+        Messages.printHorizontalLine();
+    }
+
+    public static void printFormattingError() {
+        System.out.println("Incorrectly formatted input");
+        Messages.printHorizontalLine();
+    }
+
+    public static void printNonExistError() {
+        System.out.println("Task does not exist in list");
+        Messages.printHorizontalLine();
+    }
+}
```
I like these refactorings which make the code quite clean.

Artemis-Hunt commented on src/main/java/Messages.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public static void printAddedEvent(String title) {
+        Messages.printHorizontalLine();
```
Consider removing the Messages. prefix, as it is not needed for a method in the same class

NeilBaner replied:
Good point. Will do.

Artemis-Hunt reviewed (commented):
Good job

NeilBaner committed:
Replace print() with toString()

Initially forgot about toString() existing. New implementation overrides the Object.toString() so printing can be handled by the main method.

NeilBaner committed:
Refactor input parsing methods

Refactor the complex methods to parse the input string into separate methods to improve readability.

NeilBaner committed:
Create packages for duke

Split all the various classes into packages for better organisation.

NeilBaner committed:
Reorganise some packaging and import statements

NeilBaner committed:
Add first custom exception

Add a new type of exception (UnknownCommandException) to be thrown when Duke receives a command he doesn't know how to handle.

NeilBaner committed:
Add new printable error and help messages

Add multiple new printable error and help messages, to be used when Duke throws an exception, or otherwise (including adding a help command later).

NeilBaner committed:
Add TaskIndexOutOfBoundsException

Add an exception to be thrown when the number input to mark as done doesn't exist in the list.

NeilBaner committed:
Add DukeException abstract class

Add a new class DukeException for all other exceptions to inherit from

NeilBaner committed:
Modify existing exceptions to use DukeException

Modify the existing TaskIndexOutOfBoundsException and UnknownCommandException to extend DukeException and override the printErrorMessage() method.

NeilBaner committed:
Add 3 new exception types

NeilBaner committed:
Bug fix

NeilBaner committed:
Add more exceptions

NeilBaner committed:
Begin implementing exceptions

Begin to implement the exceptions in the main Duke class.

Artemis-Hunt commented on src/main/java/com/neilbaner/duke/Duke.java:
```diff
@@ -0,0 +1,110 @@
+
+    private static boolean parseInput(String input) {
+        boolean toContinue = true;
+        if (input.toLowerCase().equals("bye")) {
+            toContinue = false;
+            Messages.printGoodBye();
+        } else if (input.toLowerCase().equals("list")) {
+            printTaskList();
+        } else if (input.toLowerCase().startsWith("done")) {
+            try {
+                int indexToMark = getIndexToMark(input);
+                markDone(indexToMark);
+            } catch (NumberFormatException e) {
+                Messages.printFormattingError();
+            } catch (IndexOutOfBoundsException e) {
+                Messages.printNonExistError();
+            }
+        } else if (input.toLowerCase().startsWith("todo")) {
+            addToDo(input.substring(5));
+        } else if (input.toLowerCase().startsWith("deadline")) {
+            try {
+                String title = getDeadlineTitle(input);
+                String dueDate = getDeadlineDueDate(input);
+                addDeadline(title, dueDate);
+            } catch (IndexOutOfBoundsException e) {
+                Messages.printFormattingError();
+            }
+        } else if (input.startsWith("event")) {
+            try {
+                String title = getEventTitle(input);
+                String eventTime = getEventTime(input);
+                addEvent(title, eventTime);
+            } catch (IndexOutOfBoundsException e) {
+                Messages.printFormattingError();
+            }
+        }
+        return toContinue;
+    }
```
Consider wrapping the whole if-else block in a try...catch and then use case/if-else to handle the various exceptions. This reduces repetitive try..catch blocks for the same error.

Artemis-Hunt commented on src/main/java/com/neilbaner/duke/Duke.java:
```diff
@@ -0,0 +1,110 @@
+
+    private static boolean parseInput(String input) {
+        boolean toContinue = true;
+        if (input.toLowerCase().equals("bye")) {
```
Consider creating String lowerCaseInput = input.toLowerCase() to reduce toLowerCase() method calls

NeilBaner replied:
Good idea, thanks.

Artemis-Hunt commented on src/main/java/com/neilbaner/duke/Duke.java:
```diff
@@ -0,0 +1,110 @@
+        Messages.printAddedDeadline(title);
+    }
+
+    public static void markDone(int indexToMark) throws NumberFormatException, IndexOutOfBoundsException {
```
Consider renaming to markAsDone

NeilBaner replied:
Sounds logical. Thanks.

Artemis-Hunt commented on src/main/java/com/neilbaner/duke/task/TaskList.java:
```diff
@@ -0,0 +1,35 @@
+        tasks = new ArrayList<Task>();
+    }
+
+    public ArrayList<Task> getAllTasks() {
```
Consider renaming to getTaskArray or getTaskList, to make it clear that the return of type ArrayList

NeilBaner replied:
Done. Thank you.

NeilBaner committed:
Merge branch 'branch-Level-5'

NeilBaner committed:
Make task abstract

NeilBaner committed:
Implement changes as per comments on PR.

NeilBaner committed:
Create lists of messages and names to call the user

NeilBaner committed:
Create a class to generate random messages based on the text files

NeilBaner committed:
Update messages class to use the randomly generated messages

NeilBaner committed:
Update content of messages and names

NeilBaner committed:
Merge branch 'branch-RandomMessages'

NeilBaner committed:
Add method to delete task and add command

NeilBaner committed:
Add message for when a task is deleted

NeilBaner committed:
Reformat some code

NeilBaner committed:
Add functions to serialize and deserialize tasks

NeilBaner committed:
Add a new type of exception to handle errors while deserializing

NeilBaner committed:
Reformat some code

NeilBaner committed:
Reformat code, set correct access modifiers

NeilBaner committed:
Add more getter methods

NeilBaner committed:
Merge branch 'master' into branch-Level-7

NeilBaner committed:
Add methods to serialize and deserialize whole task list, new delimiter

NeilBaner committed:
Add loadTask() method

NeilBaner committed:
Implement file loading and saving

NeilBaner committed:
Merge branch 'branch-Level-6'

NeilBaner committed:
Merge branch 'branch-Level-7'

NeilBaner committed:
Remove some debugging print statements

NeilBaner committed:
Fix bug in delete

NeilBaner committed:
Add javadoc comments

rishi12438 commented on src/main/java/com/neilbaner/duke/Duke.java:
```diff
@@ -0,0 +1,174 @@
+                Messages.printFormattingError();
+            }
+        } else {
+            throw new UnknownCommandException();
```
you might want to look at the situation when the user enters "    valid command"

NeilBaner replied:
I'm sorry, but I'm not entirely sure what you mean; the output I get is "I'm sorry, I'm not sure what valid command means" followed by the help text (as expected). Do you mean that the sentence itself doesn't make sense and I should enclose the user input in double quotes to make it more clear?

rishi12438 replied:
hi Neil, I mean something like "   list". The spaces infront of the command might not be recognized by your system.

rishi12438 commented on src/main/java/com/neilbaner/duke/exceptions/IncorrectFormattingException.java:
```diff
@@ -0,0 +1,12 @@
+    }
+
+    @Override
+    public void printErrorMessage(String command) {
```
why is this empty?

rishi12438 commented on src/main/java/com/neilbaner/duke/messages/MessageGenerator.java:
```diff
@@ -0,0 +1,39 @@
+public class MessageGenerator {
+    public static String generateConfirmationMessage() {
+        File confirmationMessagesFile = new File("src/main/java/com/neilbaner/duke/messages/confirmationmessages");
+        File namesFile = new File("src/main/java/com/neilbaner/duke/messages/names");
```
please try to use relative filepath .

rishi12438 commented on src/main/java/com/neilbaner/duke/serialize/Serializer.java:
```diff
@@ -0,0 +1,121 @@
+        Task deserialized;
+        parts = s.split("~");
+        try {
+            if (parts[0].equals("T")) {
```
please try to have a variable name for each of the array element so that its easier when you refer to it in the future.

rishi12438 commented on src/main/java/com/neilbaner/duke/serialize/Serializer.java:
```diff
@@ -0,0 +1,121 @@
+        ArrayList<Task> taskList = new ArrayList<Task>();
+        String[] tasks = s.split("~~");
+        for (int i = 0; i < tasks.length; i++) {
+            taskList.add(deserializeTask(tasks[i]));
```
please try to do error handling

rishi12438 commented on src/main/java/com/neilbaner/duke/Duke.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private static boolean parseInput(String input) throws DukeException {
+        String lowerCaseInput = input.toLowerCase();
```
good job in converting to lowercase!

rishi12438 reviewed (commented):
good job!

NeilBaner committed:
Update Serializer to use new delimiters, removed magic literals

NeilBaner committed:
Update saveState() to use separate lines from input

NeilBaner committed:
Update gitignore

NeilBaner committed:
Add search function in TaskList

NeilBaner committed:
Update search to not be case-sensitive anymore

NeilBaner committed:
Implement find function

NeilBaner committed:
Add new superclass with date and time

This superclass is to be used for Task types which have an associated date and time, i.e. Deadline and Event.

NeilBaner committed:
Update Event and Deadline to extend TaskWithAssociatedDate

Modify the classes to extend the new class, and also update certain methods to use the new names/fields.

NeilBaner committed:
Update Serializer

Update the Serializer class to use the newly renamed fields

NeilBaner committed:
Add BadDateFormatException

This exception is to be thrown when there's an error parsing the given date and time

NeilBaner committed:
Update README.md (WIP)

NeilBaner committed:
Change TaskWithAssociatedDate to abstract

NeilBaner committed:
Add getters and setters for TaskWithAssociatedDate

NeilBaner committed:
Update constructor

NeilBaner committed:
Correct some access modifiers

NeilBaner committed:
Override toString

NeilBaner committed:
Add functions to get tasks before and on a date

NeilBaner committed:
Change toString() to print formatted dates if possible.

NeilBaner committed:
Update to use correct method name from superclass

NeilBaner committed:
Fix formatting

NeilBaner committed:
Add methods to get tasks before and on a date

NeilBaner committed:
Merge pull request #1 from NeilBaner/branch-Level-8

Add proper date support

NeilBaner committed:
Merge branch 'master' into branch-Level-9

NeilBaner committed:
Merge pull request #2 from NeilBaner/branch-Level-9

Add a find function

NeilBaner committed:
Fix some weirdly broken code from the merge

NeilBaner committed:
Add a missing function from the merge

NeilBaner committed:
Add class to parse the input commands and store the commands

NeilBaner committed:
Add classes for file storage and relevant exceptions

NeilBaner committed:
Implement new classes in main Duke class

NeilBaner committed:
Remove magic literals and save as constants in Commands

NeilBaner committed:
Adding UI class (WIP)

NeilBaner committed:
Delete some duplicated code

NeilBaner committed:
Refactor Messages class

Reorganize the behemoth of a Messages class into a parent class and two subclasses.

NeilBaner committed:
Finish UI class

Finished making a class to handle UI interactions.

NeilBaner committed:
Create InputParser class

Create class InputParser to handle input parsing.

NeilBaner committed:
Add help command

NeilBaner committed:
Handle some more exceptions

NeilBaner committed:
Cleanup, include a backup option.

NeilBaner committed:
Update main class to take advantage of all this OOP

NeilBaner committed:
Create and update all exceptions

NeilBaner committed:
Clean up code

Various code style and quality improvements, including spacing, making certain variables final etc.

NeilBaner committed:
Update access modifiers

Update certain access modifiers to make sure that we only grant as much access as absolutely necessary,

NeilBaner committed:
Add javadoc comments for all Exceptions

NeilBaner committed:
Add javadoc comments to Storage class.

NeilBaner committed:
Fix improperly handled exception

NeilBaner committed:
Correct error

NeilBaner committed:
Correct error message

NeilBaner committed:
Update MessageGenerator to use resources

This allows it to work in a jar

NeilBaner committed:
Merge remote-tracking branch 'origin/master'

NeilBaner committed:
Merge branch 'master' into A-Javadoc

NeilBaner committed:
Add author and version tags

NeilBaner committed:
Update author and version number

NeilBaner committed:
Update gitignore

NeilBaner committed:
Add javadoc comments on all Messages classes

NeilBaner committed:
Update README.md user guide

NeilBaner committed:
Set theme jekyll-theme-hacker

NeilBaner committed:
Add more javadoc comments for package ui

NeilBaner committed:
Fix typo

NeilBaner committed:
Add javadoc comments to tasks

NeilBaner committed:
Merge branch 'A-Javadoc'

NeilBaner committed:
Fix bug found by Colin

Add additional checks for null before searching

NeilBaner committed:
Update gitignore

NeilBaner committed:
Add manifest

NeilBaner committed:
Update help command to include all the new commands.

NeilBaner committed:
Update README.md

NeilBaner committed:
Update README.md
